apps/tap/test/tap.js

```diff
@@ -62,64 +85,328 @@ contract('Tap app', accounts => {
+          it('it should revert', async () => {
+            const token = await TokenMock.new(authorized, 10000)
+            await assertRevert(
+              async () =>
+                await tap.removeTokenTap(token.address, {
+                  from: authorized
+                })
+            )
+          })
+        })
+      })
+      context('but token is not ETH or ERC20', () => {
+        it('it should revert', async () => {
+          await assertRevert(
+            async () =>
+              await tap.removeTokenTap(root, { from: authorized })
+          )
+        })
+      })
+    })
+    context('sender does not have REMOVE_TOKEN_TAP_ROLE', () => {
+      it('it should revert', async () => {
+        const token = await TokenMock.new(authorized, 10000)
+        await tap.addTokenTap(token.address, tapAmt, { from: authorized })
+
+        await assertRevert(
+          async () =>
+            await tap.removeTokenTap(token.address, { from: unauthorized })
+        )
+      })
+    })
+  })
+
+  context("updateTokenTap", () => {
+    context("sender does have UPDATE_TOKEN_TAP_ROLE", () => {
+      context('and token is ETH or ERC20', () => {
+        it("it should update tap rate if within monthly limit", async () => {
+          const token = await TokenMock.new(authorized, 10000)
+          await tap.addTokenTap(ETH, tapAmt, { from: authorized })
+          await tap.addTokenTap(token.address, tapAmt, { from: authorized })
+
+          await tap.updateTokenTap(ETH, 10, { from: authorized })
+          await tap.updateTokenTap(token.address, 10, { from: authorized })
+         // assertEvent(receipt, 'UpdateTokenTap')
+
+        })
+        it("it should revert if tap rate exceeds limit within 30 days", async () => {
+          //I'd like to submit a transaction with a spoofed date?
+          const token = await TokenMock.new(authorized, 10000)
+          await tap.addTokenTap(ETH, tapAmt, { from: authorized })
+          await tap.addTokenTap(token.address, tapAmt, { from: authorized })
+
+          /* await assertRevert(
+            async () =>
+              await tap.updateTokenTap(token.address, 100, { from: authorized })
+          ) */
+        })
+        it("it should revert if not in tap list", async () => {
+          const token = await TokenMock.new(authorized, 10000)
+          await assertRevert(
+            async () =>
+              await tap.updateTokenTap(token.address, 30, { from: authorized })
+          )
+        })
+      })
+    })
+  })
+
+  context("updateVault", () => {
+    context("sender does have UPDATE_VAULT_ROLE", () => {
+      it("it should update vault address", async () => {
+        const vault2 = await Vault.new()
+        const receipt = await tap.updateVault(vault2.address, { from: authorized })
+        assertEvent(receipt, 'UpdateVault')
+        //How do I access contract data and assert the vault address has been updated?
```

Comment:
For both Vault / Pool I'd like to access the contract memory and assert if the address has been updated. Is this correct or should I get the info from the event emission?